### ISSUE TITLE
Support for mapping multiple case classes

### ISSUE THREAD

**Issue by nsutcliffe** [scalalandio/chimney#115]
Presuming I had the following:

```scala
case class Container1(a: Int)
case class Container2(b: String)
case class MergedContainers(a: Int, b: String)
```

It would be nice to do something like either of the following:
```scala
(Container1(1),Container2("HI")).transformInto[MergedContainers]
```

Or perhaps:
```scala
case class NestedContainers(container1: Container1, container2: Container2)
NestedContainers(Container1(1),Container2("HI")).transformInto[MergedContainers]
```

**Comment by krzemin** [comment]
Having case class that combines containers (`NestedContainer`), this should be also achievable by first proposal: https://github.com/scalalandio/chimney/issues/100#issuecomment-480563946

**Comment by ittaiz** [comment]
Do you think this will be implemented sometime soon? Had a need for it a few times and used hacks but I have a feeling I’ll have a blocker in the next month or two and this could really come in handy

**Comment by krzemin** [comment]
@ittaiz many people ask for this feature recently - I think it should be next top priority. But it requires a lot of time to design it correctly, so I can't guarantee any fixed time frame.

**Comment by ittaiz** [comment]
Sure thing, I complete understand.
I appreciate the intent and thank you for your work!

**Comment by michaelbilow** [comment]
Ah, yeah, this is the feature I've been looking for. I imagine it's on the hard side to implement. Many thanks for your work on chimney @MateuszKubuszok

**Comment by MateuszKubuszok** [comment]
I have _some_ idea how to do it, but it will be a significant amount of work.

I was thinking about prioritizing 1.0.0-RC first, to not delay it anymore, and then( later in 2024) starting to work on such merging transformation (it would allow us to internally reimplement `Patcher`s as such transformer). The challenges for this feature are that:

 * it would have to merge values _recursively_ because that's where lies its true usefulness
 * it would have to allows merging and transforming an arbitrary number of values
 * it would have to cooperate with all the existing overrides (`.withField*`, `.enable*`)
 * it would have to preserve the semantics of existing code, and make sure that existing tests will still pass

I have some an idea how to do it, but it would require refactoring a few sensitive pieces of code, writing a bit more pieces of a new code, and creating a tons of tests to avoid [feature interaction bugs](https://www.hillelwayne.com/post/feature-interaction/), because at this point it's really easy to write code which might look okayish, pass a simple test, compile (obviously) but provide an unexpected behavior when combined with some other feature.

So I am _planning_ to work on it somewhere in 2024, but I have no idea how much time I'll have so I am not committing to any ETA.

**Comment by michaelbilow** [comment]
Yeah, I had gotten stuck on this issue about 4 years ago IIRC, and returned to the same problem recently. Below is my best effort to make something like this work today (leveraging some tools I'm sure you'd rather not add to `chimney`, and losing some of chimney's cleverness about failed transforms).


```scala
import io.scalaland.chimney.dsl._
import org.scalacheck.ScalacheckShapeless._
import org.scalacheck.Arbitrary


case class X(a: Int)
case class Y(b: Int)
case class Z(a: Int, b: Int)

val x = X(1)
val y = Y(2)
val z = Arbitrary.arbitrary[Z]
  .sample
  .get
  .patchUsing(x)
  .patchUsing(y)
```

**Comment by sercanersoy** [comment]
Would be great to have this feature, exactly what I'm looking for. Any updates on this?

**Comment by MateuszKubuszok** [comment]
Not really. To implement it we'd have to:
* update the DSL to allow having several source values - we cannot just use tuples, because `(a, b, c).transformInto[CaseClass]` is already reserved for [a tuple transformation](https://chimney.readthedocs.io/en/stable/supported-transformations/#frominto-a-tuple)
  * to keep the API backward compatible probably something like `foo.into[Bar].withFallbackValue(baz).transform` would be needed, where 1 value is the "main" one, and unbounded number of fallbacks is provided
  * to keep it actually useful it would have to merge values recursively - that would also allows rewriting `Patcher`s as a special case of `patch.into[Obj].withFallbackValue(obj).transform`
* then we'd update the macro to store several source values
* then update the macro to pass down the config storing several inputs, with extracting the right values recursively (so that e.g. `(foo, bar)` into `bar` would be able to figure out `(foo.a, baz.a) into `bar.a` etc)
* additionally some decisions, probably using new flags, would have to be provided for collections: should they be merged, or should the first matching source be used

All of that would have to be made in such a way that it would not break with special cases handling for:
 * Options
 * Collections
 * Eithers
 * enums
 * AnyVals
 * DSL for modifying nested overrides (e.g. `withFieldConst(_.a.matchingSubtype[A1].list.everyItem.bOpt.matchingSome, ...)`)

which means that it would be probably 1-2 weeks of full-time development of PoC, next 2 for writing unit tests for every single case when things might get difficult, then another week or two for fixing errors.

I don't see the capacity (nor feel the urge) anytime soon to sink in another 1-2 month of free full time work.

**Comment by ahoy-jon** [comment]
I was about to open a ticket for "just that". Thank you so much for doing Chimney!

**Comment by MateuszKubuszok** [comment]
The functionality is available on master, see [testing-snapshots](https://github.com/scalalandio/chimney/blob/master/CONTRIBUTING.md#testing-snapshots) if you want to help us testing it (and checking if [the documentation](https://chimney.readthedocs.io/en/latest/supported-transformations/#merging-multiple-input-sources-into-a-single-target-value) makes sense) before we'll release it (I want to do a few _much_ smaller things before the release).
